WalletWasabi/WabiSabi/Models/Decomposition/Decomposition.cs

```diff
@@ -0,0 +1,83 @@
+using System;
+using System.Linq;
+using System.Collections.Immutable;
+using NBitcoin;
+using System.Collections.Generic;
+
+namespace WalletWasabi.WabiSabi.Models.Decomposition
+{
+	public sealed record Decomposition : IComparable<Decomposition>, IEquatable<Decomposition>
+	{
+		// Useful for prettifying test assertions, or alternatively as a debug
+		// display attribute:
+		// public override string ToString()
+		// 	=> $"[ {TotalValue} = { string.Join(' ', Outputs) } ]";
+
+		internal Decomposition(params long[] outputValues)
+			: this(outputValues.Select(Money.Satoshis))
+		{
+		}
+
+		internal Decomposition(IEnumerable<Money> outputValues)
+		{
+			Outputs = outputValues.OrderByDescending(x => x).ToImmutableArray();
+			TotalValue = Outputs.Sum();
+		}
+
+		public ImmutableArray<Money> Outputs { get; private init; }
+
+		public Money TotalValue { get; private init; }
+
+		public Decomposition Extend(Money output) =>
```

Comment:
Why `Extend`? Why not just `Add`? And then you figure out where the output fits here and not delegate it to the consumer of the API. Sure, I can see it's an immutable array, but no reason to be so.

Reply:
Because the point is not to sort an array of outputs but to avoid generating the same decompositions under a different order, hence `InvalidOperationExtension`.

This should only be called from `CombinationsOfASize`, and if the concept wasn't controversial I would have written it as an inner class, which would render this additional check unnecessary.